src/ucp/stream/stream_recv.c

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2017.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include <ucp/core/ucp_ep.h>
+#include <ucp/core/ucp_worker.h>
+#include <ucp/core/ucp_context.h>
+#include <ucp/core/ucp_request.h>
+
+#include <ucs/datastruct/mpool.inl>
+#include <ucs/debug/profile.h>
+
+#include "../tag/eager.h"
+
+
+typedef struct {
+    uint64_t    uuid;
+} UCS_S_PACKED ucp_stream_eager_hdr_t;
+
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_data_nb,
+                 (ep, length), ucp_ep_h ep, size_t *length)
+{
+    ucp_recv_desc_t  *rdesc;
+    ucs_status_ptr_t ret;
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+
+    if (ucs_list_is_empty(&ep->stream_data)) {
+        ret = UCS_STATUS_PTR(UCS_OK);
+        goto out;
+    }
+
+    rdesc = ucs_list_next(&ep->stream_data, ucp_recv_desc_t,
+                          list[UCP_RDESC_HASH_LIST]);
```

Comment:
UCP_RDESC_HASH_LIST looks a bit confusing now. Maybe it is worth to think about some better name which would be common for both TAG and STREAM APIs

Reply:
HASH is pretty common name :) i would be very appreciate you would propose better one!

Reply:
@brminich is right, list[] is intended for descriptors with tags. need to use union for stream/tag data